### ISSUE TITLE
openmediavault-diskstats: salt collectd configuration mishandling boot disk

### ISSUE THREAD

**Issue by dsm1212** [openmediavault/openmediavault#1474]
So I mentioned this in another issue, but was unrelated there so I'll report separately. omv-salt generates this colllectd disk.conf on my system:

LoadPlugin disk
<Plugin disk>
	Disk "sdb"
	Disk "sdd"
	Disk "md2"
	Disk "disk/by-uuid/32f0aa31-3240-4e9d-a527-7fdd14ff140f"
	IgnoreSelected false
</Plugin>

The last entry is for my boot disk (btrfs raid 1). This string instead of md0 causes rrd to put data into subfolders of /var/lib/openmediavault/rrd. Those sub-folders don't exist so rrd errors every couple minutes. I created them and the errors go away, but they are not actually used to display results in omv.  If I edit the disk.conf and change that boot disk to md0 then the errors go away and the md0 data files are updated in the rrd folder and displayed in OMV. But when salt runs it restores the bad disk.conf file.

In December did a couple things to my boot disk that changed its UUID and I'm sure that triggered this. I can see it coincides with that point in time by the last date on the md0 captured rrd files.

I tried to wade through the salt config scripts but I don't quite see what is wrong when it regenerates this disk.conf. I did get to the point where findmnt would be called and it returns /dev/md0. Later it converts to the UUID path as it's first choice and this has one so it uses it. At least that's what I seem to be seeing when I run commands manually in python. 

My boot disk is md0, at / of course, and it has a by-name, by-label, and by-uuid entry in /dev/disk. My openmediavault config.xml has this entry for the boot disk:

      <mntent>
        <uuid>79684322-3eac-11ea-a974-63a080abab18</uuid>
        <fsname>/dev/md0</fsname>
        <dir>/</dir>
        <type>btrfs</type>
        <opts>noatime,nodiratime,defaults</opts>
        <freq>0</freq>
        <passno>0</passno>
        <hidden>1</hidden>
        <usagewarnthreshold>85</usagewarnthreshold>
        <comment></comment>
      </mntent>

I don't know what that 796xx UUID is. It's not the disk UUID, but the other mntent UUID's are also not the disk UUID's for the working ones so that UUID must be something else?

Is there a way to trace the salt run or something to debug this?

**Comment by votdev** [comment]
I'm confused and after reading the issue several times i still do not understand it.

> The last entry is for my boot disk (btrfs raid 1). This string instead of md0 causes rrd to put data into subfolders

BTRFS devices do not have anything to do with /dev/md0 device files. md0 indicates that it is a MDADM software RAID.

> The last entry is for my boot disk (btrfs raid 1).

> My boot disk is md0, at / of course

What is your boot disk now?

**Comment by dsm1212** [comment]
Gah, I was misleading, sorry. Is md raid1 but I installed btrfs. I know is uncommon but it does have a few benefits. Maybe that odd case is leading to this. I did see the code was special casing btrfs to get the device name, but later when it has the uuid the code did not have similar code to find the md0 name again. Was this recently changed maybe? Might not have anything to do with my boot disk changes…

**Comment by votdev** [comment]
Please post the output of
```
# blkid
# mount
# omv-confdbadm read --prettify conf.system.filesystem.mountpoint
```

**Comment by dsm1212** [comment]
Here you go. Thx

# blkid
/dev/sdc1: UUID="96e1785b-7085-6227-5534-c7a04888fda0" UUID_SUB="499d71c6-2ec4-9bc2-356d-a879bd38868f" LABEL="warehouse21:0" TYPE="linux_raid_member" PARTUUID="4277af2e-01"
/dev/sdc5: UUID="038af736-eba0-dd95-89b6-824a5c7b92d2" UUID_SUB="9dadc743-0a4d-7c0a-9ba2-a3d4a3ae7faf" LABEL="warehouse21:127" TYPE="linux_raid_member" PARTUUID="4277af2e-05"
/dev/sdc6: UUID="ad55e1a4-e435-382c-eced-b5f026762e64" UUID_SUB="0fcbada5-640c-ff11-aa78-69233b370314" LABEL="warehouse21:2" TYPE="linux_raid_member" PARTUUID="4277af2e-06"
/dev/sda1: UUID="96e1785b-7085-6227-5534-c7a04888fda0" UUID_SUB="fb04a062-9cbd-9701-3907-acf970905039" LABEL="warehouse21:0" TYPE="linux_raid_member" PARTUUID="4277af2e-01"
/dev/sda5: UUID="038af736-eba0-dd95-89b6-824a5c7b92d2" UUID_SUB="89a654cc-080f-6028-7ff5-dd0ca0eaf119" LABEL="warehouse21:127" TYPE="linux_raid_member" PARTUUID="4277af2e-05"
/dev/sda6: UUID="ad55e1a4-e435-382c-eced-b5f026762e64" UUID_SUB="7905a5a5-e77d-8ce5-d126-60ab655023a2" LABEL="warehouse21:2" TYPE="linux_raid_member" PARTUUID="4277af2e-06"
/dev/sdb1: LABEL="data" UUID="dcb48bcc-e271-4fd6-ba64-0f683bae3700" BLOCK_SIZE="4096" TYPE="ext4" PARTUUID="97b0d05d-5c78-4523-89b2-b14fb9bcb592"
/dev/md127: UUID="6a3b7e2d-513f-4fbf-a736-69b99fed2c52" TYPE="swap"
/dev/md2: LABEL="Home" UUID="92f79319-8123-4cf6-b14e-79d5c5c017fc" UUID_SUB="dccc4d80-7d73-40dd-b308-df232f76ad86" BLOCK_SIZE="4096" TYPE="btrfs"
/dev/md0: LABEL="ROOT" UUID="32f0aa31-3240-4e9d-a527-7fdd14ff140f" UUID_SUB="c0f4ceea-e5bd-436a-9d9b-33f02a0f5ba2" BLOCK_SIZE="4096" TYPE="btrfs"
/dev/sdd1: LABEL="BACKUPS" UUID="f52ef3a3-ad82-4f14-a194-5ad9297b214d" BLOCK_SIZE="4096" TYPE="ext4" PARTUUID="8899363a-01"

# mount
sysfs on /sys type sysfs (rw,nosuid,nodev,noexec,relatime)
proc on /proc type proc (rw,nosuid,nodev,noexec,relatime)
udev on /dev type devtmpfs (rw,nosuid,relatime,size=8007244k,nr_inodes=2001811,mode=755,inode64)
devpts on /dev/pts type devpts (rw,nosuid,noexec,relatime,gid=5,mode=620,ptmxmode=000)
tmpfs on /run type tmpfs (rw,nosuid,nodev,noexec,relatime,size=1606996k,mode=755,inode64)
/dev/md0 on / type btrfs (rw,noatime,nodiratime,ssd,space_cache,subvolid=5,subvol=/)
securityfs on /sys/kernel/security type securityfs (rw,nosuid,nodev,noexec,relatime)
tmpfs on /dev/shm type tmpfs (rw,nosuid,nodev,inode64)
tmpfs on /run/lock type tmpfs (rw,nosuid,nodev,noexec,relatime,size=5120k,inode64)
cgroup2 on /sys/fs/cgroup type cgroup2 (rw,nosuid,nodev,noexec,relatime,nsdelegate,memory_recursiveprot)
pstore on /sys/fs/pstore type pstore (rw,nosuid,nodev,noexec,relatime)
bpf on /sys/fs/bpf type bpf (rw,nosuid,nodev,noexec,relatime,mode=700)
systemd-1 on /proc/sys/fs/binfmt_misc type autofs (rw,relatime,fd=30,pgrp=1,timeout=0,minproto=5,maxproto=5,direct,pipe_ino=1717)
hugetlbfs on /dev/hugepages type hugetlbfs (rw,relatime,pagesize=2M)
mqueue on /dev/mqueue type mqueue (rw,nosuid,nodev,noexec,relatime)
debugfs on /sys/kernel/debug type debugfs (rw,nosuid,nodev,noexec,relatime)
tracefs on /sys/kernel/tracing type tracefs (rw,nosuid,nodev,noexec,relatime)
configfs on /sys/kernel/config type configfs (rw,nosuid,nodev,noexec,relatime)
fusectl on /sys/fs/fuse/connections type fusectl (rw,nosuid,nodev,noexec,relatime)
sunrpc on /run/rpc_pipefs type rpc_pipefs (rw,relatime)
tmpfs on /tmp type tmpfs (rw,relatime,inode64)
/dev/md2 on /vol/dev-disk-by-id-md-name-warehouse21-2 type btrfs (rw,relatime,ssd,space_cache,subvolid=5,subvol=/)
/dev/sdd1 on /vol/dev-disk-by-uuid-f52ef3a3-ad82-4f14-a194-5ad9297b214d type ext4 (rw,relatime,jqfmt=vfsv0,usrjquota=aquota.user,grpjquota=aquota.group)
/dev/sdb1 on /vol/dev-disk-by-uuid-dcb48bcc-e271-4fd6-ba64-0f683bae3700 type ext4 (rw,relatime,stripe=64,jqfmt=vfsv0,usrjquota=aquota.user,grpjquota=aquota.group)
/dev/md2 on /vol/dev-disk-by-id-md-name-warehouse21-2/docker/btrfs type btrfs (rw,relatime,ssd,space_cache,subvolid=5,subvol=/)
nsfs on /run/docker/netns/default type nsfs (rw)
nsfs on /run/docker/netns/c73f2c61a15f type nsfs (rw)
nsfs on /run/docker/netns/49e27cd94419 type nsfs (rw)
nsfs on /run/docker/netns/e03865444f36 type nsfs (rw)
nsfs on /run/docker/netns/6e53076fcd93 type nsfs (rw)
nsfs on /run/docker/netns/4824e609931a type nsfs (rw)
nsfs on /run/docker/netns/b36d4c3bf092 type nsfs (rw)
nsfs on /run/docker/netns/d2608cd7d9de type nsfs (rw)
nsfs on /run/docker/netns/057b525f38f2 type nsfs (rw)
nsfs on /run/docker/netns/4ddf4c0fd2a3 type nsfs (rw)
nsfs on /run/docker/netns/cfaed969bc71 type nsfs (rw)
nsfs on /run/docker/netns/53063d89c88d type nsfs (rw)
nsfs on /run/docker/netns/a35eff35cb75 type nsfs (rw)
nsfs on /run/docker/netns/5e6e590c5a3d type nsfs (rw)
nsfs on /run/docker/netns/f5185d937618 type nsfs (rw)
nsfs on /run/docker/netns/8282c3a68125 type nsfs (rw)
nsfs on /run/docker/netns/81d192182309 type nsfs (rw)
nsfs on /run/docker/netns/9694222a8a8b type nsfs (rw)
tmpfs on /run/user/1000 type tmpfs (rw,nosuid,nodev,relatime,size=1606992k,nr_inodes=401748,mode=700,uid=1000,gid=100,inode64)

# omv-confdbadm read --prettify conf.system.filesystem.mountpoint
[
    {
        "comment": "",
        "dir": "/vol/dev-disk-by-uuid-dcb48bcc-e271-4fd6-ba64-0f683bae3700",
        "freq": 0,
        "fsname": "/dev/disk/by-uuid/dcb48bcc-e271-4fd6-ba64-0f683bae3700",
        "hidden": false,
        "opts": "defaults,nofail,user_xattr,usrjquota=aquota.user,grpjquota=aquota.group,jqfmt=vfsv0,acl",
        "passno": 2,
        "type": "ext4",
        "usagewarnthreshold": 85,
        "uuid": "e9f055a2-f6df-4c5a-8304-e0303f1e1c2b"
    },
    {
        "comment": "",
        "dir": "/vol/dev-disk-by-uuid-f52ef3a3-ad82-4f14-a194-5ad9297b214d",
        "freq": 0,
        "fsname": "/dev/disk/by-uuid/f52ef3a3-ad82-4f14-a194-5ad9297b214d",
        "hidden": false,
        "opts": "defaults,nofail,user_xattr,usrjquota=aquota.user,grpjquota=aquota.group,jqfmt=vfsv0,acl",
        "passno": 2,
        "type": "ext4",
        "usagewarnthreshold": 85,
        "uuid": "404be80f-e13c-4948-acd5-52292fc97e13"
    },
    {
        "comment": "",
        "dir": "/vol/dev-disk-by-id-md-name-warehouse21-2",
        "freq": 0,
        "fsname": "/dev/disk/by-id/md-name-warehouse21:2",
        "hidden": false,
        "opts": "defaults,nofail,ssd",
        "passno": 2,
        "type": "btrfs",
        "usagewarnthreshold": 85,
        "uuid": "a6ff651d-83d9-403f-96e8-ebd4a1aa5f6c"
    },
    {
        "comment": "",
        "dir": "/",
        "freq": 0,
        "fsname": "/dev/md0",
        "hidden": true,
        "opts": "noatime,nodiratime,defaults",
        "passno": 0,
        "type": "btrfs",
        "usagewarnthreshold": 85,
        "uuid": "79684322-3eac-11ea-a974-63a080abab18"
    }
]

**Comment by votdev** [comment]
Do you have installed the `openmediavault-sharerootfs` plugin?

**Comment by votdev** [comment]
Could you please run the following python code interactively?
The commands after `>>>` must be entered by yourself per line.
```
root@omv6box:/home/vagrant# python3
Python 3.9.2 (default, Feb 28 2021, 17:03:44) 
[GCC 10.2.1 20210110] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import openmediavault.fs
>>> fs_mp = openmediavault.fs.Filesystem.from_mount_point('/')
>>> dev_name = fs_mp.predictable_device_file
>>> dev_name
'/dev/disk/by-uuid/084c29b3-cc2f-473d-aea9-93051946599b'
>>> fs.parent_device_file
'/dev/vda'
```

**Comment by dsm1212** [comment]
Yes I have sharerootfs installed.

Commands below. But I think the issue is not with these commands. When adding the root filesystem in disk.sls the calls are different. First it calls omv_util.get_root_filesystem() which returns the uuid device string. Then that is passed to omv_util.get_fs_parent_device_file() which calls these two:
    fs = openmediavault.fs.Filesystem(id_)
    return fs.parent_device_file
This last line returns the uuid path and not /dev/md0 when I invoke it in python. So the uuid path gets used.

# python3
Python 3.9.2 (default, Feb 28 2021, 17:03:44)
[GCC 10.2.1 20210110] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import openmediavault.fs
>>> fs_mp = openmediavault.fs.Filesystem.from_mount_point('/')
>>> dev_name = fs_mp.predictable_device_file
>>> dev_name
'/dev/disk/by-uuid/32f0aa31-3240-4e9d-a527-7fdd14ff140f'
>>> fs_mp.parent_device_file
'/dev/md0'


Python 3.9.2 (default, Feb 28 2021, 17:03:44)
[GCC 10.2.1 20210110] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import openmediavault.fs
>>> fs = openmediavault.fs.Filesystem.from_mount_point('/')
>>> fs.predictable_device_file
'/dev/disk/by-uuid/32f0aa31-3240-4e9d-a527-7fdd14ff140f'
>>> fs_id = openmediavault.fs.Filesystem('/dev/disk/by-uuid/32f0aa31-3240-4e9d-a527-7fdd14ff140f')
>>> fs_id.parent_device_file
'/dev/disk/by-uuid/32f0aa31-3240-4e9d-a527-7fdd14ff140f'

**Comment by votdev** [comment]
Fixed in `openmediavault-diskstats 6.0.2`.

**Comment by dsm1212** [comment]
Took the update and it worked! Thanks
